### PR TITLE
Add GIoU loss

### PR DESCRIPTION
**Pull request by quantumalaviya**
# What does this PR do?
Implements the Generalized IoU loss

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/keras-team/keras-cv/blob/master/.github/CONTRIBUTING.md),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue? Please add a link
      to it if that's the case.
- [x] Did you write any new necessary tests?

## Who can review?
@LukeWood

### CODE REVIEW

File: keras_cv/losses/giou_loss.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+        boxes1_rank = len(boxes1.shape)
+        boxes2_rank = len(boxes2.shape)
+
+        if (
+            boxes1_rank != boxes2_rank
+            or boxes1_rank not in [2, 3]
+            or boxes2_rank not in [2, 3]
+        ):
+            raise ValueError(
+                "_compute_giou() expects both boxes to be batched, or both "
```
Comment (LukeWood) [inline keras_cv/losses/giou_loss.py:71]: instead of "_compute_giou() expects" tell users "`GIoULoss` expects"
tell users the public API name in value errors - thats who the errors are for in the end of the day

File: keras_cv/losses/giou_loss.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+                f"boxes to be unbatched.  Received len(boxes1.shape)={boxes1_rank}, "
```
Comment (LukeWood) [inline keras_cv/losses/giou_loss.py:72]: backticks around code blocks.

`len(boxes1.shape=` instead of len(boxes1.shape)

File: keras_cv/losses/giou_loss.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+                f"len(boxes2.shape)={boxes2_rank}.  Expected either len(boxes1.shape)=2 AND "
+                "len(boxes2.shape)=2, or len(boxes1.shape)=3 AND len(boxes2.shape)=3."
+            )
+
+        if self.bounding_box_format.startswith("rel"):
```
Comment (LukeWood) [inline keras_cv/losses/giou_loss.py:77]: did we think about upstreaming this to the utils so that it automatically strips the rel_ prefix if they both use rel_

File: keras_cv/losses/giou_loss.py [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+        bounding_box_format: a case-insensitive string which is one of `"xyxy"`,
+            `"rel_xyxy"`, `"xyWH"`, `"center_xyWH"`, `"yxyx"`, `"rel_yxyx"`.
+            Each bounding box is defined by at least these 4 values. The inputs
+            may contain additional information such as classes and confidence after
+            these 4 values but these values will be ignored while calculating
+            this loss. For detailed information on the supported formats, see the
+            [KerasCV bounding box documentation](https://keras.io/api/keras_cv/bounding_box/formats/).
+
+    References:
+        - [GIoU paper](https://arxiv.org/pdf/1902.09630)
```
Comment (LukeWood) [inline keras_cv/losses/giou_loss.py:42]: any references to a torch or other implementation you based this on we should include?  Could be helpful in reviewing.

File: keras_cv/losses/giou_loss.py [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+            enclose_area = enclose_width * enclose_height
+            return iou - tf.math.divide_no_nan(
+                (enclose_area - union_area), enclose_area
+            )
+
+        if boxes1_rank == 2:
+            return compute_giou_for_batch((boxes1, boxes2))
+        else:
+            return tf.map_fn(
+                compute_giou_for_batch, elems=(boxes1, boxes2), dtype=boxes1.dtype
```
Comment (LukeWood) [inline keras_cv/losses/giou_loss.py:132]: can this use vectorized_map?

File: keras_cv/losses/giou_loss.py (hunk 3 above)
Comment (quantumalaviya) [reply keras_cv/losses/giou_loss.py:77]: `compute_iou` already has this. Do you mean a dedicated util function only for this?

File: keras_cv/losses/giou_loss.py (hunk 4 above)
Comment (quantumalaviya) [reply keras_cv/losses/giou_loss.py:42]: I used the [tfa implementation](https://www.tensorflow.org/addons/api_docs/python/tfa/losses/GIoULoss)

Comment (LukeWood) [reply keras_cv/losses/giou_loss.py:42]: can you add this as a reference?  Thanks!

File: keras_cv/losses/giou_loss.py [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+    GIoU loss is a modified IoU loss commonly used for object detection. This loss aims
+    to directly optimize the IoU score between true boxes and predicted boxes. GIoU loss
+    adds a penalty term to the IoU loss that takes in account the area of the
+    smallest box enclosing both the boxes being considered for the iou. The length of
+    the last dimension should be atleast 4 to represent the bounding boxes. While
+    this dimension can have more than 4 values, these values will be ignored for the
+    calculation of this loss.
+
+    Args:
+        bounding_box_format: a case-insensitive string which is one of `"xyxy"`,
```
Comment (LukeWood) [inline keras_cv/losses/giou_loss.py:33]: lets not provide an exhaustive list and instead redirect to the docs for this.  That way when we update it the docstrings get updated too.

File: keras_cv/losses/giou_loss.py (hunk 3 above)
Comment (quantumalaviya) [reply keras_cv/losses/giou_loss.py:77]: `convert_format` already strips the rel_ prefix if both use rel_ (I think this was updated in one of the PRs I had made earlier). This part is just to ensure that the rel_ is consistent in both source and target.

**Comment by quantumalaviya** [comment]
Closing this PR for now as this loss will be covered as part of the upcoming YoloX PR
